Change query to queries

gribnoysup opened the pull request:
A possible solution for https://github.com/ReactTraining/react-media/issues/69

As described in this issue:
- `query` prop is removed
- `queries` prop is added
- `queries` prop values are media queries (and not the prop itself)
- `render` and `children` props are both functions and `render` prop is an alias for children (they behave the same way)

I changed tests for browser environment and also added new tests for the server environment.

mjackson commented on modules/__tests__/Media-server-test.js:
```diff
@@ -0,0 +1,53 @@
+      ReactDOMServer.renderToStaticMarkup(
+        <Media queries={queries}>
+          {matches => {
+            expect(matches).toMatchObject(expected)
```
Let's make the assertion on the markup instead of in the render prop and change the assertion to be something like "it renders all queries".

mjackson commented on modules/__tests__/Media-server-test.js:
```diff
@@ -0,0 +1,53 @@
+      ReactDOMServer.renderToStaticMarkup(
+        <Media queries={queries} defaultMatches={defaultMatches}>
+          {matches => {
+            expect(matches).toMatchObject(defaultMatches)
```
Same as above.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -24,52 +26,85 @@ describe('A <Media>', () => {
+        ReactDOM.render(
+          <Media queries={queries}>
+            {matches => {
+              expect(matches).toMatchObject(expected)
```
Let's test the output of render instead of the arguments to the render prop, just so we have complete coverage, and change the description to be something like "it renders its children function" or similar.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -24,52 +26,85 @@ describe('A <Media>', () => {
+      it('should call children with expected values', () => {
+        ReactDOM.render(
+          <Media queries={queries} render={matches => {
+            expect(matches).toMatchObject(expected)
```
Same as above.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -79,67 +114,80 @@ describe('A <Media>', () => {
+        ReactDOM.render(
+          <Media queries={queries}>
+            {matches => {
+              expect(matches).toMatchObject(expected)
```
Same as above.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -79,67 +114,80 @@ describe('A <Media>', () => {
+      it('should call children with expected values', () => {
+        ReactDOM.render(
+          <Media queries={queries} render={matches => {
+            expect(matches).toMatchObject(expected)
```
Same as above.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -79,67 +114,80 @@ describe('A <Media>', () => {
+        ReactDOM.render(
+          <Media queries={queries}>
+            {matches => {
+              expect(matches).toMatchObject(expected)
```
Same as above.

mjackson commented on modules/__tests__/Media-test.js:
```diff
@@ -79,67 +114,80 @@ describe('A <Media>', () => {
+      it('should call children with expected values', () => {
+        ReactDOM.render(
+          <Media queries={queries} render={matches => {
+            expect(matches).toMatchObject(expected)
```
Same as above.

mjackson commented on modules/Media.js:
```diff
@@ -2,72 +2,70 @@ import React from 'react'
     const { children, render } = this.props
     const { matches } = this.state
 
-    return (
```
Let's make changes to the rendering logic in a separate PR, please. As written, this change eliminates support for non-function `children`.

gribnoysup replied:
Sure! I think I just didn't understand correctly what you meant in the comment :)

gribnoysup commented:
Hi @mjackson 

I updated the PR, but there are some alterations, because with the `query` prop matches was just a flag, and now it is an object with multiple flags. 

I tried to recreate the behavior of the render method as close as possible to the one, that is in the master branch:

- `render` prop not called at all when there are no `matches` with the `queries` (same as before)
  - `render` now have `matches` as a first argument
- `children` component, they are not rendered at all, if there are no matches (same as before)
    - we provide `matches` as a prop only if the child is a composite component (new behavior)
- `children` function always called with matches as a first argument (same as before)

The tests are also updated

gribnoysup commented:
Resolved merge conflicts and prettified everything

edorivai commented:
@gribnoysup I feel like this was closed accidentally, feel free to close again if it was really closed on purpose. @mjackson do you still feel like releasing this as v2? Looking at #69 there's quite some interest in the feature!

gribnoysup commented:
Hi @edorivai! I didn't see any updates for the PR after I made all the changes so I assumed that it is not needed anymore. TBH rebasing on prettier changes was hard enough the first time, so if this PR is needed I would rather just do it from scratch

gribnoysup commented:
Done, the branch is updated. Hope it wouldn't go stale again

edorivai commented:
@gribnoysup thanks a lot! I decided to merge into the `next` branch for now, as these changes are breaking and we'll probably want to release it as a major version bump.

@mjackson How would you want to move forward from here? Perhaps cut a v2 alpha release?

Newnab commented:
What's the current status of this? Is it still in the works?

I have a use case where lots of props on a component (in my case, a datepicker) are changed depending on which media queries match, and it would be preferable to me to have a way to have multiple queries in one rather than declaring the component multiple times inside separate <Media> components, if possible!

edorivai commented:
Hey @Newnab :wave:

The status is that we're waiting for @mjackson to jump in and publish the changes to npm. It seems that Michael is a bit busy lately, so I'm not sure when he'll take a look here.
